tsdb/shard.go

```diff
@@ -1736,73 +1732,127 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
 	return len(fs.fields) == 0
 }
 
-func (fs *MeasurementFieldSet) Save() (err error) {
-	// current version
-	var v uint64
-	// Is the MeasurementFieldSet empty?
-	isEmpty := false
-	// marshaled MeasurementFieldSet
+type errorChannel chan<- error
 
-	b, err := func() ([]byte, error) {
-		fs.mu.Lock()
-		defer fs.mu.Unlock()
-		fs.memoryVersion += 1
-		v = fs.memoryVersion
-		// If no fields left, remove the fields index file
-		if len(fs.fields) == 0 {
-			isEmpty = true
-			if err := os.RemoveAll(fs.path); err != nil {
-				return nil, err
-			} else {
-				fs.writtenVersion = fs.memoryVersion
-				return nil, nil
-			}
-		}
-		return fs.marshalMeasurementFieldSetNoLock()
-	}()
+type MeasurementFieldSetWriter struct {
+	writeErrChannels chan errorChannel
+}
 
-	if err != nil {
-		return err
-	} else if isEmpty {
-		return nil
-	}
+// SetMeasurementFieldSetWriter - initialize the queue for write requests
+// and start the background write process
+func (fs *MeasurementFieldSet) SetMeasurementFieldSetWriter(queueLength int) {
+	fs.writer = &MeasurementFieldSetWriter{writeErrChannels: make(chan errorChannel, queueLength)}
+	go fs.saveWriter()
+}
 
-	// Write the new index to a temp file and rename when it's sync'd
-	// if it is still the most recent memoryVersion of the MeasurementFields
-	path := fs.path + "." + strconv.FormatUint(v, 10) + ".tmp"
+func (fs *MeasurementFieldSet) Save() error {
+	return fs.writer.RequestSave()
+}
 
-	fd, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_EXCL|os.O_SYNC, 0666)
-	if err != nil {
-		return err
-	}
-	defer os.RemoveAll(path)
+func (w *MeasurementFieldSetWriter) RequestSave() error {
+	done := make(chan error)
+	w.writeErrChannels <- done
```

Comment:
Read from a newly created channel not used elsewhere should be blocking forever?

Reply:
The channel is the one used to signal the completion of the write to file, and should be used exactly once.

Reply:
Ah sorry, I misinterpreted a a read from done rather than a send to writeErrChannels.

Can we wrap this in a `request` struct? Sending a channel in a channel is too confusing.

Reply:
Sure.  You had had that in your prototype, and I didn't like the proliferation of single-field structures. But, if you think it's clearer, happy to oblige.